### PR TITLE
Vb 2110 update existing sessions to use the new enhanced type method

### PR DESCRIPTION
**Pull request by dhirajshettymoj**
## What does this pull request do?

Updates the incentive logic to use grouping instead of a boolean.
## What is the intent behind these changes?

To bring the sessions for incentive levels logic in line with category and housing locations.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/sessions/CreateSessionTemplateDto.kt [hunk 1]
```diff
@@ -51,9 +51,6 @@ data class CreateSessionTemplateDto(
   @Schema(description = "list of group references for permitted session location groups", required = false)
   val locationGroupReferences: List<String>? = listOf(),
 
-  @Schema(description = "enhanced privilege", example = "true", required = true)
-  val enhanced: Boolean,
-
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/sessions/CreateSessionTemplateDto.kt:56]: dont we need this?

Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/sessions/CreateSessionTemplateDto.kt:56]: Updated to include incentiveLevelGroupReferences.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SessionService.kt [hunk 2]
```diff
@@ -108,31 +108,22 @@ class SessionService(
...
       return sessionTemplates.filter { sessionTemplate ->
-        sessionTemplate.permittedSessionCategoryGroups.isEmpty() || isPrisonerCategoryAllowedOnSession(sessionTemplate, prisonerCategory)
+        sessionValidator.isSessionAvailableToPrisonerCategory(prisonerCategory, sessionTemplate)
       }
     }
 
     return sessionTemplates
   }
 
-  fun isPrisonerCategoryAllowedOnSession(sessionTemplate: SessionTemplate, prisonerCategory: String?): Boolean {
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SessionService.kt:118]: where has this gone?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/PrisonerCategoryMatcher.kt [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Component
+import uk.gov.justice.digital.hmpps.visitscheduler.model.entity.session.SessionTemplate
+import java.util.function.BiPredicate
+import java.util.stream.Collectors
+
+@Component
+class PrisonerCategoryMatcher : BiPredicate<String?, SessionTemplate> {
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/PrisonerCategoryMatcher.kt:11]: nice

File: src/main/resources/session-template-data/template.ftl [hunk 4]
```diff
@@ -184,10 +187,77 @@
...
+    -- update tmp incentive group table with correct prison id for given code.
+    UPDATE tmp_session_incentive_group SET prison_id = prison.id FROM prison WHERE tmp_session_incentive_group.prison_code = prison.code;
+
+    -- insert data into real session incentive group table from temporary one.
+    INSERT INTO session_incentive_group(id,reference,prison_id,name)
+    SELECT id,CONCAT('-',REGEXP_REPLACE(to_hex((ROW_NUMBER () OVER (ORDER BY key))+2951597050), '(.{3})(?!$)', '\1~','g')) as reference,prison_id,name FROM tmp_session_incentive_group order by id;
+
+    -- Sequence updated manually as id's were inserted from temp table
+    ALTER SEQUENCE session_incentive_group_id_seq RESTART WITH  ${session_incentive_group_id_index};
+
```
Comment (AledWEvans) [inline src/main/resources/session-template-data/template.ftl:219]: awesome

**Review by AledWEvans — APPROVED** [review]
LGTM, but wonder if you should do VB-2216 before you merge this in!?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SessionService.kt (hunk 2 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SessionService.kt:118]: moved this into PrisonerSessionValidator

**Pull request merged**